### PR TITLE
AVX-62741: add a simple cache for the app-domains API

### PR DESCRIPTION
**Pull request by jpeach**
Since there is no way to query the state of a single entry from the app-domains API, add a very conservative cache that mitigates refreshing a lot of smart or web groups. We deliberately serialize access the the API endpoint, dropping the cache after any modification or a conservative 5sec timeout.

### CODE REVIEW

File: goaviatrix/appdomain_cache.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
+package goaviatrix
+
+import (
```
Comment (dnardo) [inline goaviatrix/appdomain_cache.go:3]: How did we test this?  I'm just wondering if did any tests where we add entries continuously.
For instance with accounts we had something like this:

```
resource "aviatrix_account" "aws_account" {
  count               = 500
  account_name = "aws-${count.index +1}"
  cloud_type = 1
....
  ]
}
```

And we saw some strange corner cases with the TF state while doing that.  You may not have these issues, but I just want to make sure we did some sort of test with bulk updates.

I wonder also how much the cache time plays a role here also.

Comment (jpeach) [reply goaviatrix/appdomain_cache.go:3]: The field/support folks tested this with the terraform attached to AVX-60981.

**Review by dnardo — APPROVED** [review]

**Comment by kspoelstra-aviatrix** [comment]
LGTM, @jpeach , I think this is an important part of the solution, I'm not quite sure about the edge cases with terraform, do we want to make this configurable in case we run into unforeseen edgecases? 

For large deployments we are going to support using multiple terraform  in 8.1.0.  The cache should not be an issue as users should have their own smartgroups and not collide.
We also do have a change coming where we can query the smartgroups individually, taking away the large request. But still doing multiple requests and we're subject to latency there. 

I don't think we have any OOB smartgroup creation triggered by other terraform which could give us issues here . (@shinrich , do you know of any non smartgroup operations which create smartgroups? (Enabling DCF?))

**Comment by jpeach** [comment]
> do we want to make this configurable in case we run into unforeseen edgecases

We don't have a way to do that today. As you note, the right medium term fix is to improve the API :)
